atmel-samd/asf4/patches/samd21/cast-align.patch

```diff
@@ -10,3 +10,27 @@
 +        #pragma GCC diagnostic pop
  	struct timer_task *      it    = (struct timer_task *)list_get_head(&timer->tasks);
  	uint32_t                 time  = ++timer->time;
+--- a/hal/src/hal_flash.c
++++ b/hal/src/hal_flash.c
+@@ -304,7 +304,10 @@ static int32_t flash_is_address_aligned(struct flash_descriptor *flash, const ui
+  */
+ static void flash_ready(struct _flash_device *device)
+ {
++       #pragma GCC diagnostic push
++       #pragma GCC diagnostic ignored "-Wcast-align"
+        struct flash_descriptor *const descr = CONTAINER_OF(device, struct flash_descriptor, dev);
```

Comment:
Should something have been declared `COMPILER_WORD_ALIGNED` or something like that to avoid the `-Wcast-align` warning suppression here and elsewhre? I'm just trying to understand why you need them.

Reply:
I think this is an artifact of the ASF code for `CONTAINER_OF`. Their Makefile doesn't use `-Wcast-align` but ours does and I'd rather whitelist places than turn it off. Here is the error:

```
In file included from asf4/samd51/hal/src/hal_flash.c:46:0:
asf4/samd51/hal/src/hal_flash.c: In function 'flash_ready':
asf4/samd51/hal/utils/include/utils.h:60:46: error: cast increases required alignment of target type [-Werror=cast-align]
 #define CONTAINER_OF(ptr, type, field_name) ((type *)(((uint8_t *)ptr) - offsetof(type, field_name)))
                                              ^
asf4/samd51/hal/src/hal_flash.c:307:41: note: in expansion of macro 'CONTAINER_OF'
  struct flash_descriptor *const descr = CONTAINER_OF(device, struct flash_descriptor, dev);
                                         ^~~~~~~~~~~~
cc1: all warnings being treated as errors
make: *** [build-metro_m4_express/asf4/samd51/hal/src/hal_flash.o] Error 1
make: *** Waiting for unfinished jobs....
```

Reply:
Oh, bleah, OK, I'd guess it's the `uint8_t` in `CONTAINER_OF`.